Fix issue causing disk cache eviction LRU to not function as expected.

timonus opened the pull request:
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

### Pull Request Description

`NSFileManager` does not automatically increment access date when reading files off disk, which means that use of `SDImageCacheConfigExpireTypeAccessDate` is effectively identical to using `SDImageCacheConfigExpireTypeCreationDate`. This doesn't seem like the correct behavior on SDWebImage's part, and it isn't what's documented [here](https://github.com/SDWebImage/SDWebImage/blob/d68b92db8557f1aed608106e5d6ae8826ce3afac/SDWebImage/Core/SDImageCacheConfig.h#L15). This PR makes it so that SDWebImage adjusts the file content access date when files are read off disk to address the issue.

I've also added a commit that updates the default cache behavior to `SDImageCacheConfigExpireTypeAccessDate` since I believe that's what's commonly expected. People generally want cache purging by time-of-use in my experience instead of using the time the file was downloaded at. If you strongly disagree I can omit that change!

timonus commented:
I have done this in my image loading library for quite a long time (though I used to use modification date) https://github.com/timonus/TJImageCache/blob/main/TJImageCache/TJImageCache.m#L196

dreampiggy commented:
Sounds great. I'll have a check later.

Actually these code was not changed at least 5 years...😂 So it's broken at the beginning ?

dreampiggy commented on SDWebImage/Core/SDDiskCache.m:
```diff
@@ -71,13 +71,16 @@ - (NSData *)dataForKey:(NSString *)key {
     }
     NSData *data = [NSData dataWithContentsOfFile:filePath options:self.config.diskCacheReadingOptions error:nil];
     if (data) {
+        [[NSURL fileURLWithPath:filePath] setResourceValue:[NSDate date] forKey:NSURLContentAccessDateKey error:nil];
```
Is this time consuming ?

If so, need a extra `if (self.config.diskCacheExpireType == SDImageCacheConfigExpireTypeAccessDate`) guard

If not, it's OK

timonus replied:
It is quite fast, significantly faster than the data read in my (very quick) testing.

<img width="766" alt="Screenshot 2024-10-08 at 11 12 33 AM" src="https://github.com/user-attachments/assets/64737a41-1c1d-4cc1-b816-cf93c7fdf159">

Logged times are (1) time to read data, (2) time to write content access date, and (3) 2/1

dreampiggy commented on SDWebImage/Core/SDImageCacheConfig.m:
```diff
@@ -34,7 +34,7 @@ - (instancetype)init {
         _diskCacheWritingOptions = NSDataWritingAtomic;
         _maxDiskAge = kDefaultCacheMaxDiskAge;
         _maxDiskSize = 0;
-        _diskCacheExpireType = SDImageCacheConfigExpireTypeModificationDate;
```
It's ok to change default value, but this PR will marked as `Behavior Changes` and I'll point this on release changelog

Actually I don't know why the original author use `write` and expire type, since it's not actually LRU (we usually don't write to same cache key image)

dreampiggy commented:
Can you rebase and run the test cases again (or update the test case with your changes, since you change the default value)

timonus commented:
Alright, rebased and linting seems to pass. It looks like unit tests are all failing to build but I don't think that's caused by the contents of this PR. I updated the one test that referenced modified date [here](https://github.com/SDWebImage/SDWebImage/pull/3759/files#diff-492e89c523b525ff6fc7120dab615ca0d993fadd4b2ad1e65b2b1c357f9311bd).

timonus commented:
@dreampiggy anything I can do to nudge this along? I'm unable to run the tests locally without installing Cocoapods, which I need to make other changes for and I'd rather not do. I don't think the tests are failing due to my change but if they are could you let me know?

dreampiggy commented:
The `SDWebImageTestCache.m` has compile error, seems I can fix this

dreampiggy commented on Tests/Tests/SDWebImageTestCache.m:
```diff
@@ -91,7 +91,7 @@ - (void)removeDataForKey:(nonnull NSString *)key {
     NSDate *expirationDate = [NSDate dateWithTimeIntervalSinceNow:-self.config.maxDiskAge];
     NSURL *diskCacheURL = [NSURL fileURLWithPath:self.cachePath isDirectory:YES];
-    NSArray<NSString *> *resourceKeys = @[NSURLIsDirectoryKey, NSURLAttributeModificationDateKey];
+    NSArray<NSString *> *resourceKeys = @[NSURLIsDirectoryKey, NSURLAttributeContentAccessDateKey];
```
```suggestion
    NSArray<NSString *> *resourceKeys = @[NSURLIsDirectoryKey, NSURLContentAccessDateKey];
```

dreampiggy commented on Tests/Tests/SDWebImageTestCache.m:
```diff
@@ -108,8 +108,8 @@ - (void)removeExpiredData {
             // Remove files that are older than the expiration date;
-            NSDate *modifiedDate = resourceValues[NSURLAttributeModificationDateKey];
-            if (expirationDate && [[modifiedDate laterDate:expirationDate] isEqualToDate:expirationDate]) {
+            NSDate *accessDate = resourceValues[NSURLAttributeContentAccessDateKey];
```
```suggestion
            NSDate *accessDate = resourceValues[NSURLContentAccessDateKey];
```

timonus commented:
Ah, thank you! Sorry I didn't catch that.

dreampiggy commented:
Merge for now. But the release date of 5.20 may be delayed due to other break changes I want to added.
